test: scheduling: add test case

yingmingx opened the pull request:
Add a test case of preemptive thread scheduling.
The scheduler will select the highest priority and
waiting longest thread to be the current thread.

Signed-off-by: Ying ming <mingx.ying@intel.com>

jenmwms commented:
It looks good to me, but where can I see results? Shippable states the run was successful but I cannot view any of the test output when I navigate there. I wanted to confirm the new test `test_priority_preemptible_wait_prio` is selected by sanitycheck as part of `sched_api` testsuite in main.c. It would be good to see successful test on multiple archs.

jenmwms commented:
> It looks good to me, but where can I see results? ...

nvm. I downloaded the log and looked through it. Looks good, except I see a skipped test due to SRAM overflow. Is that due to this patch perhaps? 

- INFO    -  78/146 lpcxpresso55s16_ns        tests/kernel/sched/schedule_api/kernel.scheduler   [33mSKIPPED[39m (SRAM overflow)

Also, I wonder why the large difference in runtime for some of the tests. Since this patch supports multiple configurations for the testsuite, some configurations will use more time to complete so this is not an issue here. Please correct me if I misunderstood.
INFO    - 135/146 qemu_xtensa               tests/kernel/sched/schedule_api/kernel.scheduler   [32mPASSED[39m (qemu 13.687s)

- INFO    - 139/146 qemu_nios2                tests/kernel/sched/schedule_api/kernel.scheduler   [32mPASSED[39m (qemu 29.751s)

- INFO    - 144/146 qemu_x86                  tests/kernel/sched/schedule_api/kernel.scheduler   [32mPASSED[39m (qemu 9.354s)

nashif commented:
> It looks good to me, but where can I see results?

you should run those locally for closer inspection, shippable is not suitable for reviewing details results.

jenmwms commented:
> > It looks good to me, but where can I see results?
> 
> you should run those locally for closer inspection, shippable is not suitable for reviewing details results.

Thank you! Ok, I pulled this code and confirmed locally this passes for  `qemu_x86` , `qemu_x86_64` , and `qemu_xtensa` as part of sched_api testsuite using `west build -p -b build qemu_x86 tests/kernel/sched/schedule_api` and `west build -t run`

stephanosio commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +109,58 @@ void test_priority_preemptible(void)
+	tid[0] = k_thread_create(&tdata_prio[0], tstacks[0], STACK_SIZE,
+			thread_entry_prio, INT_TO_POINTER(0), NULL, NULL,
+			K_PRIO_PREEMPT(1), 0, K_NO_WAIT);
+	/*the highest-priority thread that has waited the longest*/
```
```suggestion
	/* the highest-priority thread that has waited the longest */
```

stephanosio commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +109,58 @@ void test_priority_preemptible(void)
+	tid[1] = k_thread_create(&tdata_prio[1], tstacks[1], STACK_SIZE,
+			thread_entry_prio, INT_TO_POINTER(1), NULL, NULL,
+			K_PRIO_PREEMPT(0), 0, K_MSEC(10));
+	/*the highest-priority thread that has waited shorter*/
```
```suggestion
	/* the highest-priority thread that has waited shorter */
```

stephanosio commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +109,58 @@ void test_priority_preemptible(void)
+
+	zassert_true(tid_num == 1, NULL);
+
+	/*test case tear down*/
```
```suggestion
	/* test case tear down */
```

stephanosio commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +109,58 @@ void test_priority_preemptible(void)
+/**
+ * @brief Validate preemptiveness of preemptive thread
+ *
+ * @details Create three preemptive thread which is of priority
+ * higher than current thread. Make sure that the highest priority
+ * and longest waiting thread is scheduled.
```
```suggestion
 * @details Create three preemptive threads which are of priority
 * higher than current thread. Make sure that the highest priority
 * and longest waiting thread is scheduled first.
```

stephanosio commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +109,58 @@ void test_priority_preemptible(void)
 }
 
+/**
+ * @brief Validate preemptiveness of preemptive thread
```
```suggestion
 * @brief Validate scheduling sequence of preemptive threads with start delay
```

yingmingx replied:
Thank you, that helps very much.

stephanosio reviewed (commented):
Nits

stephanosio reviewed (commented):
Just a suggestion: this test, in its current form, only validates that the longest waiting thread with the highest priority is scheduled first and does not validate subsequent scheduled threads.

I wonder if we can make this test more comprehensive by:

1. Validating that `K_PRIO_PREEMPT(0), K_MSEC(20)` is indeed scheduled after `K_PRIO_PREEMPT(0), K_MSEC(10)`
2. Validating that `K_PRIO_PREEMPT(1), K_NO_WAIT` is scheduled after `K_PRIO_PREEMPT(0), K_MSEC(20)`
3. Replicating the delay mechanism for `K_PRIO_PREEMPT(1)` such that the full validation, at minimum, covers:
    ```
    { K_PRIO_PREEMPT(0), K_MSEC(10) } ->
    { K_PRIO_PREEMPT(0), K_MSEC(20) } ->
    { K_PRIO_PREEMPT(1), K_MSEC(10) } ->
    { K_PRIO_PREEMPT(1), K_MSEC(20) }
    ```

yingmingx commented:
> Just a suggestion: this test, in its current form, only validates that the longest waiting thread with the highest priority is scheduled first and does not validate subsequent scheduled threads.
> 
> I wonder if we can make this test more comprehensive by:
> 
> 1. Validating that `K_PRIO_PREEMPT(0), K_MSEC(20)` is indeed scheduled after `K_PRIO_PREEMPT(0), K_MSEC(10)`
> 2. Validating that `K_PRIO_PREEMPT(1), K_NO_WAIT` is scheduled after `K_PRIO_PREEMPT(0), K_MSEC(20)`
> 3. Replicating the delay mechanism for `K_PRIO_PREEMPT(1)` such that the full validation, at minimum, covers:
>    ```
>    { K_PRIO_PREEMPT(0), K_MSEC(10) } ->
>    { K_PRIO_PREEMPT(0), K_MSEC(20) } ->
>    { K_PRIO_PREEMPT(1), K_MSEC(10) } ->
>    { K_PRIO_PREEMPT(1), K_MSEC(20) }
>    ```

yingmingx commented:
> Just a suggestion: this test, in its current form, only validates that the longest waiting thread with the highest priority is scheduled first and does not validate subsequent scheduled threads.
> 
> I wonder if we can make this test more comprehensive by:
> 
> 1. Validating that `K_PRIO_PREEMPT(0), K_MSEC(20)` is indeed scheduled after `K_PRIO_PREEMPT(0), K_MSEC(10)`
> 2. Validating that `K_PRIO_PREEMPT(1), K_NO_WAIT` is scheduled after `K_PRIO_PREEMPT(0), K_MSEC(20)`
> 3. Replicating the delay mechanism for `K_PRIO_PREEMPT(1)` such that the full validation, at minimum, covers:
>    ```
>    { K_PRIO_PREEMPT(0), K_MSEC(10) } ->
>    { K_PRIO_PREEMPT(0), K_MSEC(20) } ->
>    { K_PRIO_PREEMPT(1), K_MSEC(10) } ->
>    { K_PRIO_PREEMPT(1), K_MSEC(20) }
>    ```

Thank you , your suggestion is helpful. I have changed the code according to your suggestion. But I found there are some error
on platform qemu_arc, so I issued a bug below.
https://github.com/zephyrproject-rtos/zephyr/issues/26500

nashif commented:
You need to rebase this PR

andyross commented on tests/kernel/sched/schedule_api/prj.conf:
```diff
@@ -2,6 +2,6 @@ CONFIG_ZTEST=y
 CONFIG_NUM_PREEMPT_PRIORITIES=30
 #CONFIG_SCHED_SCALABLE=y
-CONFIG_MAX_THREAD_BYTES=4
+CONFIG_MAX_THREAD_BYTES=6
```
This looks suspicious.  You aren't creating any new `struct k_thread` objects, just re-using ones that already exist.  So the total number of thread kobj indexes shouldn't be changing.  If it is, you're probably re-using a thread that hasn't aborted yet (though IIRC that should be detected and fail), or there's a bug somewhere else.  Why do you need this setting?

yingmingx replied:
In this testcase, I want to creat 4 thread like following
{ K_PRIO_PREEMPT(0), K_MSEC(10) } ->
{ K_PRIO_PREEMPT(0), K_MSEC(20) } ->
{ K_PRIO_PREEMPT(1), K_MSEC(10) } ->
{ K_PRIO_PREEMPT(1), K_MSEC(20) }
the sanitycheck would have a compile mistake and the build.log file told me to change the macro CONFIG_MAX_THREAD_BYTES value, it should be 5 not 6.

andyross replied:
OK, we need to figure this out.  @andrewboie is the expert here.  But my read of the code is that the kobj indexes for those objects should have been freed for reuse by their previous thread aborts.  If that didn't happen, we either have a bug where the test is re-using threads incorrectly or thread abort isn't doing the right object free.

andrewboie replied:
if you are getting a **compiler** error, this isn't a bug with re-use of thread indexes, or indeed a runtime issue at all.

what is happening is that the entire zephyr kernel binary is searched for `struct k_thread` objects, and the ones found at build time get an index assigned to them, up to `CONFIG_MAX_THREAD_BYTES * 8`

so if you get a a build error that CONFIG_MAX_THREAD_BYTES is too small, then what has happened is that the number of `struct k_thread` objects found by the build exceeds the number of bits available defined by CONFIG_MAX_THREAD_BYTES.

This change would imply that a value of 4 (32 thread objects) was not enough. You can see what objects the build has found by running make with `VERBOSE=1`.

It does look like this test creates a lot of objects, without this PR the verbose output is:

```
gen_kobject_list.py: symbol 'tstacks' at 0x12c000 contains 10 object(s)
gen_kobject_list.py: symbol 'sema2' at 0x1890d0 contains 1 object(s)
gen_kobject_list.py: symbol 'sema3' at 0x1890e8 contains 1 object(s)
gen_kobject_list.py: symbol 't' at 0x124000 contains 10 object(s)
gen_kobject_list.py: symbol 'tdata' at 0x124600 contains 1 object(s)
gen_kobject_list.py: symbol 'end_sema' at 0x12546c contains 1 object(s)
gen_kobject_list.py: symbol 'tdata' at 0x1246a0 contains 1 object(s)
gen_kobject_list.py: symbol 'tthread' at 0x124820 contains 3 object(s)
gen_kobject_list.py: symbol 't' at 0x124780 contains 1 object(s)
gen_kobject_list.py: symbol 'pend_sema' at 0x189088 contains 1 object(s)
gen_kobject_list.py: symbol 'timer_sema' at 0x189100 contains 1 object(s)
gen_kobject_list.py: symbol 'timer' at 0x124740 contains 1 object(s)
gen_kobject_list.py: symbol 'sema' at 0x1890a0 contains 1 object(s)
gen_kobject_list.py: DIE DW_TAG_variable, size=14, has_children=False
    |DW_AT_name        :  AttributeValue(name='DW_AT_name', form='DW_FORM_string', value=b't', raw_value=b't', offset=49715)
    |DW_AT_decl_file   :  AttributeValue(name='DW_AT_decl_file', form='DW_FORM_data1', value=7, raw_value=7, offset=49717)
    |DW_AT_decl_line   :  AttributeValue(name='DW_AT_decl_line', form='DW_FORM_data1', value=124, raw_value=124, offset=49718)
    |DW_AT_decl_column :  AttributeValue(name='DW_AT_decl_column', form='DW_FORM_data1', value=18, raw_value=18, offset=49719)
    |DW_AT_type        :  AttributeValue(name='DW_AT_type', form='DW_FORM_ref4', value=7095, raw_value=7095, offset=49720)
    |DW_AT_location    :  AttributeValue(name='DW_AT_location', form='DW_FORM_exprloc', value=[145, 164, 124], raw_value=[145, 164, 124], offset=49724)

gen_kobject_list.py: File '/home/apboie/projects/zephyr/zephyr/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c', line 124:
gen_kobject_list.py:     kernel object 't' found on stack
gen_kobject_list.py: symbol 't' at 0x124a40 contains 10 object(s)
gen_kobject_list.py: symbol 'sema1' at 0x1890b8 contains 1 object(s)
gen_kobject_list.py: symbol 'user_thread' at 0x125040 contains 1 object(s)
gen_kobject_list.py: symbol 'user_sem' at 0x189118 contains 1 object(s)
gen_kobject_list.py: symbol 'ustack' at 0x168000 contains 1 object(s)
gen_kobject_list.py: symbol 'fdtable_lock' at 0x189030 contains 1 object(s)
gen_kobject_list.py: symbol '__device_sys_init_z_clock_driver_init0' at 0x189020 contains 1 object(s)
gen_kobject_list.py: symbol 'ztest_thread' at 0x125240 contains 1 object(s)
gen_kobject_list.py: symbol 'cpuhold_sem' at 0x1254d4 contains 1 object(s)
gen_kobject_list.py: symbol 'ztest_thread_stack' at 0x16e000 contains 1 object(s)
gen_kobject_list.py: symbol 'uart_ns16550_driver_api' at 0x1109dc contains 1 object(s)
gen_kobject_list.py: symbol '__device_uart_ns16550_0' at 0x189010 contains 1 object(s)
gen_kobject_list.py: symbol '__device_uart_ns16550_1' at 0x189000 contains 1 object(s)
gen_kobject_list.py: symbol 'z_main_stack' at 0x17a000 contains 1 object(s)
gen_kobject_list.py: symbol 'z_main_thread' at 0x125380 contains 1 object(s)
gen_kobject_list.py: symbol 'z_idle_threads' at 0x1252e0 contains 1 object(s)
gen_kobject_list.py: symbol 'z_idle_stacks' at 0x174000 contains 1 object(s)
gen_kobject_list.py: symbol 'z_interrupt_stacks' at 0x180000 contains 1 object(s)
gen_kobject_list.py: DIE DW_TAG_variable, size=17, has_children=False
    |DW_AT_name        :  AttributeValue(name='DW_AT_name', form='DW_FORM_strp', value=b'dummy_thread', raw_value=16052, offset=302356)
    |DW_AT_decl_file   :  AttributeValue(name='DW_AT_decl_file', form='DW_FORM_data1', value=1, raw_value=1, offset=302360)
    |DW_AT_decl_line   :  AttributeValue(name='DW_AT_decl_line', form='DW_FORM_data2', value=458, raw_value=458, offset=302361)
    |DW_AT_decl_column :  AttributeValue(name='DW_AT_decl_column', form='DW_FORM_data1', value=18, raw_value=18, offset=302363)
    |DW_AT_type        :  AttributeValue(name='DW_AT_type', form='DW_FORM_ref4', value=796, raw_value=796, offset=302364)
    |DW_AT_location    :  AttributeValue(name='DW_AT_location', form='DW_FORM_exprloc', value=[145, 220, 126], raw_value=[145, 220, 126], offset=302368)

gen_kobject_list.py: File '/home/apboie/projects/zephyr/zephyr/kernel/init.c', line 458:
gen_kobject_list.py:     kernel object 'dummy_thread' found on stack
gen_kobject_list.py: symbol 'offload_sem' at 0x189070 contains 1 object(s)
gen_kobject_list.py: symbol 'pipe_async_msgs' at 0x18904c contains 1 object(s)
gen_kobject_list.py: device instance at 0x189020 has no associated subsystem
gen_kobject_list.py: found 60 kernel object instances total
```

@andyross at least in the current version of this PR, some new static thread objects are introduced in test_thread_priority.c:

```
static struct k_thread tdata_prio[THREAD_NUM];
```

If we were right at the limit beforehand, this could push the necessary amount of thread bytes from 4 to 5 (which the PR currently does, don't know why it was increased to 6 in an earlier version)

tl;dr: I think this is OK.

yingmingx replied:
@andyross Hi, andyross.  It seems that andrewboie has made it clear why the system prompts to modify the configuration file.

andyross commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +111,63 @@ void test_priority_preemptible(void)
+	/* set current thread to a non-negative priority */
+	last_prio = 2;
+	k_thread_priority_set(k_current_get(), last_prio);
+
```
This is a little racy.  You don't know that a millisecond boundary will not happen between these thread creation calls, throwing your expected math off.  Put a `k_usleep(1)` call here (or anything else that will block on a timeout), that will ensure that this code starts running at the start of a system tick and minimize the probability of a timing glitch.

Alternatively, there is now a facility for absolute/uptime timeouts (e.g `K_TIMEOUT_ABS_*()`) which you can use instead and which will guarantee to fire at the time you expect without the possibility of race conditions vs. real time.

yingmingx replied:
Thank you for your valuable advice.

LeiW000 replied:
@yingmingx , have you addressed Andy's comment?

yingmingx replied:
@andyross . Hi, Andy, I have addressed this comment. Please take a review, thanks.

andyross commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -95,6 +111,63 @@ void test_priority_preemptible(void)
+	k_sem_init(&sync_sema, 0, THREAD_NUM);
+
+	/* set current thread to a non-negative priority */
+	last_prio = 2;
```
Style: here, you're using a fixed number ("2") for a thread priority.  Below, you're using the (IMHO confusing -- preemptible and cooperative priorities count in different directions!) K_PRIO_* macros instead. 

Pick one convention.  I suggest this one.

yingmingx replied:
I will unify it.

andyross reviewed (changes requested):
The test looks OK with a few nitpicks.

The userspace thread index exhaustion is creeping me out though and it needs a clearer answer.

tyrantlizard77 reviewed (commented):
thread index change looks OK to me. PR generally looks good but please address any outstanding comments by @andyross

Commit:
test: scheduling: add test case

Add a test case of preemptive thread scheduling.
The scheduler will select the highest priority and
waiting longest thread to be the current thread.

Signed-off-by: Ying ming <mingx.ying@intel.com>

yingmingx commented:
@andyross, all the comment have been addressed. Please take a look at it, thanks.
